File: storage/ndb/src/ndbapi/NdbOperation.cpp
/*
   Copyright (c) 2003, 2024, Oracle and/or its affiliates.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License, version 2.0,
   as published by the Free Software Foundation.

   This program is designed to work with certain software (including
   but not limited to OpenSSL) that is licensed under separate terms,
   as designated in a particular file or component or in included license
   documentation.  The authors of MySQL hereby grant you an additional
   permission to link the program and your derivative works with the
   separately licensed software that they have either included with
   the program or referenced in the documentation.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License, version 2.0, for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA
*/

#include <signaldata/TcKeyReq.hpp>
#include "API.hpp"

/******************************************************************************
 * NdbOperation(Ndb* aNdb, Table* aTable);
 *
 * Return Value:  None
 * Parameters:    aNdb: Pointers to the Ndb object.
 *                aTable: Pointers to the Table object
 * Remark:        Create an object of NdbOperation.
 ****************************************************************************/
NdbOperation::NdbOperation(Ndb *aNdb, NdbOperation::Type aType)
    : m_type(aType),
      theReceiver(aNdb),
      theErrorLine(0),
      theNdb(aNdb),
      // theTable(aTable),
      theNdbCon(nullptr),
      theNext(nullptr),
      theTCREQ(nullptr),
      theFirstATTRINFO(nullptr),
      theCurrentATTRINFO(nullptr),
      theTotalCurrAI_Len(0),
      theAI_LenInCurrAI(0),
      theLastKEYINFO(nullptr),

      theFirstLabel(nullptr),
      theLastLabel(nullptr),
      theFirstBranch(nullptr),
      theLastBranch(nullptr),
      theFirstCall(nullptr),
      theLastCall(nullptr),
      theFirstSubroutine(nullptr),
      theLastSubroutine(nullptr),
      theNoOfLabels(0),
      theNoOfSubroutines(0),

      m_currentTable(nullptr),  // theTableId(0xFFFF),
      m_accessTable(nullptr),   // theAccessTableId(0xFFFF),
      // theSchemaVersion(0),
      theTotalNrOfKeyWordInSignal(8),
      theTupKeyLen(0),
      theNoOfTupKeyLeft(0),
      theOperationType(NotDefined),
      theStatus(Init),
      theMagicNumber(0xFE11D0),
      theScanInfo(0),
      m_tcReqGSN(GSN_TCKEYREQ),
      m_keyInfoGSN(GSN_KEYINFO),
      m_attrInfoGSN(GSN_ATTRINFO),
      theBlobList(nullptr),
      m_abortOption(-1),
      m_noErrorPropagation(false),
      theLockHandle(nullptr),
      m_blob_lock_upgraded(false) {
  theReceiver.init(NdbReceiver::NDB_OPERATION, this);
  theError.code = 0;
  m_customData = nullptr;
}
/*****************************************************************************
 * ~NdbOperation();
 *
 * Remark:         Delete tables for connection pointers (id).
 *****************************************************************************/
NdbOperation::~NdbOperation() {
  assert(theRequest == nullptr);  // The same as theTCREQ
}
/******************************************************************************
 *void setErrorCode(int anErrorCode);
 *
 * Remark:         Set an Error Code on operation and
 *                 on connection set an error status.
 *****************************************************************************/
void NdbOperation::setErrorCode(int anErrorCode) const {
  /* Setting an error is considered to be a const
     operation, hence the nasty cast here */
  NdbOperation *pnonConstThis = const_cast<NdbOperation *>(this);

  pnonConstThis->theError.code = anErrorCode;
  theNdbCon->theErrorLine = theErrorLine;
  theNdbCon->theErrorOperation = pnonConstThis;
  if (!(m_abortOption == AO_IgnoreError && m_noErrorPropagation))
    theNdbCon->setOperationErrorCode(anErrorCode);
}

/******************************************************************************
 * void setErrorCodeAbort(int anErrorCode);
 *
 * Remark:         Set an Error Code on operation and on connection set
 *                 an error status.
 *****************************************************************************/
void NdbOperation::setErrorCodeAbort(int anErrorCode) const {
  /* Setting an error is considered to be a const
     operation, hence the nasty cast here */
  NdbOperation *pnonConstThis = const_cast<NdbOperation *>(this);

  pnonConstThis->theError.code = anErrorCode;
  theNdbCon->theErrorLine = theErrorLine;
  theNdbCon->theErrorOperation = pnonConstThis;
  // ignore m_noErrorPropagation
  theNdbCon->setOperationErrorCodeAbort(anErrorCode);
}

/*****************************************************************************
 * int init();
 *
 * Return Value:  Return 0 : init was successful.
 *                Return -1: In all other case.
 * Remark:        Initiates operation record after allocation.
 *****************************************************************************/

int NdbOperation::init(const NdbTableImpl *tab, NdbTransaction *myConnection) {
  theStatus = Init;
  theError.code = 0;
  theErrorLine = 1;
  m_currentTable = m_accessTable = tab;

  theNdbCon = myConnection;
  for (Uint32 i = 0; i < NDB_MAX_NO_OF_ATTRIBUTES_IN_KEY; i++)
    for (int j = 0; j < 3; j++) theTupleKeyDefined[i][j] = 0;

  theFirstATTRINFO = nullptr;
  theCurrentATTRINFO = nullptr;
  theLastKEYINFO = nullptr;

  theTupKeyLen = 0;
  theNoOfTupKeyLeft = tab->getNoOfPrimaryKeys();

  theTotalCurrAI_Len = 0;
  theAI_LenInCurrAI = 0;
  theStartIndicator = 0;
  theCommitIndicator = 0;
  theSimpleIndicator = 0;
  theDirtyIndicator = 0;
  theReadCommittedBaseIndicator = 0;
  theInterpretIndicator = 0;
  theDistrKeyIndicator_ = 0;
  theScanInfo = 0;
  theTotalNrOfKeyWordInSignal = 8;
  theMagicNumber = getMagicNumber();
  m_attribute_record = nullptr;
  theBlobList = nullptr;
  m_abortOption = -1;
  m_noErrorPropagation = false;
  m_flags = 0;
  m_flags |= OF_NO_DISK;
  m_interpreted_code = nullptr;
  m_extraSetValues = nullptr;
  m_numExtraSetValues = 0;
  m_customData = nullptr;

  if (theReceiver.init(NdbReceiver::NDB_OPERATION, this)) {
    // theReceiver sets the error code of its owner
    return -1;
  }

  NdbApiSignal *tSignal = theNdb->getSignal();
  if (tSignal == nullptr) {
    setErrorCode(4000);
    return -1;
  }
  theTCREQ = tSignal;
  theTCREQ->setSignal(m_tcReqGSN, refToBlock(theNdbCon->m_tcRef));

  theAI_LenInCurrAI = 20;
  TcKeyReq *const tcKeyReq = CAST_PTR(TcKeyReq, theTCREQ->getDataPtrSend());
  tcKeyReq->scanInfo = 0;
  theKEYINFOptr = &tcKeyReq->keyInfo[0];
  theATTRINFOptr = &tcKeyReq->attrInfo[0];

  if (theNdb->theImpl->get_ndbapi_config_parameters().m_default_queue_option)
    m_flags |= OF_QUEUEABLE;

  return 0;
}

/******************************************************************************
 * void release();
 *
 * Remark:        Release all objects connected to the operation object.
 *****************************************************************************/
void NdbOperation::release() {
  NdbBlob *tBlob;
  NdbBlob *tSaveBlob;

  /* In case we didn't execute... */
  postExecuteRelease();

  tBlob = theBlobList;
  while (tBlob != nullptr) {
    tSaveBlob = tBlob;
    tBlob = tBlob->theNext;
    theNdb->releaseNdbBlob(tSaveBlob);
  }
  theBlobList = nullptr;
  theReceiver.release();

  theLockHandle = nullptr;
  m_blob_lock_upgraded = false;

#ifndef NDEBUG
  // Poison members to detect late usage
  m_accessTable = m_currentTable = (NdbTableImpl *)0x1;
  theNdbCon = (NdbTransaction *)0x1;
  m_key_record = m_attribute_record = (NdbRecord *)0x1;
#endif
}

void NdbOperation::postExecuteRelease() {
  NdbApiSignal *tSignal;
  NdbApiSignal *tSaveSignal;
  NdbBranch *tBranch;
  NdbBranch *tSaveBranch;
  NdbLabel *tLabel;
  NdbLabel *tSaveLabel;
  NdbCall *tCall;
  NdbCall *tSaveCall;
  NdbSubroutine *tSubroutine;
  NdbSubroutine *tSaveSubroutine;

  tSignal = theRequest; /* TCKEYREQ/TCINDXREQ/SCANTABREQ */
  while (tSignal != nullptr) {
    tSaveSignal = tSignal;
    tSignal = tSignal->next();
    theNdb->releaseSignal(tSaveSignal);
  }
  theRequest = nullptr;
  theLastKEYINFO = nullptr;
#ifdef TODO
  /**
   * Compute correct #cnt signals between theFirstATTRINFO/theCurrentATTRINFO
   */
  if (theFirstATTRINFO) {
    theNdb->releaseSignals(1, theFirstATTRINFO, theCurrentATTRINFO);
  }
#else
  tSignal = theFirstATTRINFO;
  while (tSignal != nullptr) {
    tSaveSignal = tSignal;
    tSignal = tSignal->next();
    theNdb->releaseSignal(tSaveSignal);
  }
#endif
  theFirstATTRINFO = nullptr;
  theCurrentATTRINFO = nullptr;

  if (theInterpretIndicator == 1) {
    tBranch = theFirstBranch;
    while (tBranch != nullptr) {
      tSaveBranch = tBranch;
      tBranch = tBranch->theNext;
      theNdb->releaseNdbBranch(tSaveBranch);
    }
    tLabel = theFirstLabel;
    while (tLabel != nullptr) {
      tSaveLabel = tLabel;
      tLabel = tLabel->theNext;
      theNdb->releaseNdbLabel(tSaveLabel);
    }
    tCall = theFirstCall;
    while (tCall != nullptr) {
      tSaveCall = tCall;
      tCall = tCall->theNext;
      theNdb->releaseNdbCall(tSaveCall);
    }
    tSubroutine = theFirstSubroutine;
    while (tSubroutine != nullptr) {
      tSaveSubroutine = tSubroutine;
      tSubroutine = tSubroutine->theNext;
      theNdb->releaseNdbSubroutine(tSaveSubroutine);
    }
  }
}

NdbRecAttr *NdbOperation::getValue(const char *anAttrName, char *aValue) {
  return getValue_impl(m_currentTable->getColumn(anAttrName), aValue);
}

NdbRecAttr *NdbOperation::getValue(Uint32 anAttrId, char *aValue) {
  return getValue_impl(m_currentTable->getColumn(anAttrId), aValue);
}

NdbRecAttr *NdbOperation::getValue(const NdbDictionary::Column *col,
                                   char *aValue) {
  if (theStatus != UseNdbRecord)
    return getValue_impl(&NdbColumnImpl::getImpl(*col), aValue);

  setErrorCodeAbort(4508);
  /* GetValue not allowed for NdbRecord defined operation */
  return nullptr;
}

int NdbOperation::equal(const char *anAttrName, const char *aValuePassed) {
  const NdbColumnImpl *col = m_accessTable->getColumn(anAttrName);
  if (col == nullptr) {
    setErrorCode(4004);
    return -1;
  } else {
    return equal_impl(col, aValuePassed);
  }
}

int NdbOperation::equal(Uint32 anAttrId, const char *aValuePassed) {
  const NdbColumnImpl *col = m_accessTable->getColumn(anAttrId);
  if (col == nullptr) {
    setErrorCode(4004);
    return -1;
  } else {
    return equal_impl(col, aValuePassed);
  }
}

int NdbOperation::setValue(const char *anAttrName, const char *aValuePassed) {
  const NdbColumnImpl *col = m_currentTable->getColumn(anAttrName);
  if (col == nullptr) {
    setErrorCode(4004);
    return -1;
  } else {
    return setValue(col, aValuePassed);
  }
}

int NdbOperation::setValue(Uint32 anAttrId, const char *aValuePassed) {
  const NdbColumnImpl *col = m_currentTable->getColumn(anAttrId);
  if (col == nullptr) {
    setErrorCode(4004);
    return -1;
  } else {
    return setValue(col, aValuePassed);
  }
}

NdbBlob *NdbOperation::getBlobHandle(const char *anAttrName) {
  // semantics differs from overloaded 'getBlobHandle(const char*) const'
  // by delegating to the non-const variant of internal getBlobHandle(...),
  // which may create a new BlobHandle
  const NdbColumnImpl *col = m_currentTable->getColumn(anAttrName);
  if (col == nullptr) {
    setErrorCode(4004);
    return nullptr;
  } else {
    return getBlobHandle(theNdbCon, col);
  }
}

NdbBlob *NdbOperation::getBlobHandle(Uint32 anAttrId) {
  // semantics differs from overloaded 'getBlobHandle(Uint32) const'
  // by delegating to the non-const variant of internal getBlobHandle(...),
  // which may create a new BlobHandle
  const NdbColumnImpl *col = m_currentTable->getColumn(anAttrId);
  if (col == nullptr) {
    setErrorCode(4004);
    return nullptr;
  } else {
    return getBlobHandle(theNdbCon, col);
  }
}

NdbBlob *NdbOperation::getBlobHandle(const char *anAttrName) const {
  const NdbColumnImpl *col = m_currentTable->getColumn(anAttrName);
  if (col == nullptr) {
    setErrorCode(4004);
    return nullptr;
  } else {
    return getBlobHandle(theNdbCon, col);
  }
}

NdbBlob *NdbOperation::getBlobHandle(Uint32 anAttrId) const {
  const NdbColumnImpl *col = m_currentTable->getColumn(anAttrId);
  if (col == nullptr) {
    setErrorCode(4004);
    return nullptr;
  } else {
    return getBlobHandle(theNdbCon, col);
  }
}

int NdbOperation::incValue(const char *anAttrName, Uint32 aValue) {
  return incValue(m_currentTable->getColumn(anAttrName), aValue);
}

int NdbOperation::incValue(const char *anAttrName, Uint64 aValue) {
  return incValue(m_currentTable->getColumn(anAttrName), aValue);
}

int NdbOperation::incValue(Uint32 anAttrId, Uint32 aValue) {
  return incValue(m_currentTable->getColumn(anAttrId), aValue);
}

int NdbOperation::incValue(Uint32 anAttrId, Uint64 aValue) {
  return incValue(m_currentTable->getColumn(anAttrId), aValue);
}

int NdbOperation::subValue(const char *anAttrName, Uint32 aValue) {
  return subValue(m_currentTable->getColumn(anAttrName), aValue);
}

int NdbOperation::subValue(const char *anAttrName, Uint64 aValue) {
  return subValue(m_currentTable->getColumn(anAttrName), aValue);
}

int NdbOperation::subValue(Uint32 anAttrId, Uint32 aValue) {
  return subValue(m_currentTable->getColumn(anAttrId), aValue);
}

int NdbOperation::subValue(Uint32 anAttrId, Uint64 aValue) {
  return subValue(m_currentTable->getColumn(anAttrId), aValue);
}

int NdbOperation::read_attr(const char *anAttrName, Uint32 RegDest) {
  return read_attr(m_currentTable->getColumn(anAttrName), RegDest);
}

int NdbOperation::read_attr(Uint32 anAttrId, Uint32 RegDest) {
  return read_attr(m_currentTable->getColumn(anAttrId), RegDest);
}

int NdbOperation::write_attr(const char *anAttrName, Uint32 RegDest) {
  return write_attr(m_currentTable->getColumn(anAttrName), RegDest);
}

int NdbOperation::write_attr(Uint32 anAttrId, Uint32 RegDest) {
  return write_attr(m_currentTable->getColumn(anAttrId), RegDest);
}

const char *NdbOperation::getTableName() const {
  return m_currentTable->m_externalName.c_str();
}

const NdbDictionary::Table *NdbOperation::getTable() const {
  return m_currentTable;
}

NdbTransaction *NdbOperation::getNdbTransaction() const { return theNdbCon; }

int NdbOperation::getLockHandleImpl() {
  assert(!theLockHandle);

  if (likely(((theOperationType == ReadRequest) ||
              (theOperationType == ReadExclusive)) &&
             (m_type == PrimaryKeyAccess) &&
             ((theLockMode == LM_Read) | (theLockMode == LM_Exclusive)))) {
    theLockHandle = theNdbCon->getLockHandle();
    if (!theLockHandle) {
      return 4000;
    }

    /* Now operation has a LockHandle - it'll be
     * filled-in when the operation is prepared prior
     * to execution.
     */
    assert(theLockHandle->m_state == NdbLockHandle::ALLOCATED);
    assert(!theLockHandle->isLockRefValid());

    return 0;
  } else {
    /* getLockHandle only supported for primary key read with a lock */
    return 4549;
  }
}

const NdbLockHandle *NdbOperation::getLockHandle() {
  if (likely(!m_blob_lock_upgraded)) {
    if (theLockHandle == nullptr) {
      int rc = getLockHandleImpl();

      if (likely(rc == 0))
        return theLockHandle;
      else {
        setErrorCode(rc);
        return nullptr;
      }
    }
    /* Return existing LockHandle */
    return theLockHandle;
  } else {
    /* Not allowed to call getLockHandle() on a Blob-upgraded
     * read
     */
    setErrorCode(4549);
    return nullptr;
  }
}

const NdbLockHandle *NdbOperation::getLockHandle() const {
  /* NdbRecord / handle already exists variant */
  return theLockHandle;
}

File: storage/ndb/src/ndbjtie/com/mysql/ndbjtie/ndbapi/NdbOperation.java
/*
  Copyright (c) 2010, 2023, Oracle and/or its affiliates.

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License, version 2.0,
  as published by the Free Software Foundation.

  This program is also distributed with certain software (including
  but not limited to OpenSSL) that is licensed under separate terms,
  as designated in a particular file or component or in included license
  documentation.  The authors of MySQL hereby grant you an additional
  permission to link the program and your derivative works with the
  separately licensed software that they have included with MySQL.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License, version 2.0, for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA
*/
/*
 * NdbOperation.java
 */

package com.mysql.ndbjtie.ndbapi;

import java.nio.ByteBuffer;

import com.mysql.jtie.Wrapper;
import com.mysql.jtie.ArrayWrapper;

public class NdbOperation extends Wrapper implements NdbOperationConst
{
    public /*_virtual_*/ native NdbBlob/*_NdbBlob *_*/ getBlobHandle(String/*_const char *_*/ anAttrName) /*_const_*/; // MMM nameclash with non-const version
    public /*_virtual_*/ native NdbBlob/*_NdbBlob *_*/ getBlobHandle(int/*_Uint32_*/ anAttrId) /*_const_*/; // MMM nameclash with non-const version
    public final native NdbErrorConst/*_const NdbError &_*/ getNdbError() /*_const_*/;
    public final native int getNdbErrorLine() /*_const_*/;
    public final native String/*_const char *_*/ getTableName() /*_const_*/;
    public final native NdbDictionary.TableConst/*_const NdbDictionary.Table *_*/ getTable() /*_const_*/;
    public final native int/*_Type_*/ getType() /*_const_*/;
    public final native int/*_LockMode_*/ getLockMode() /*_const_*/;
    public final native int/*_AbortOption_*/ getAbortOption() /*_const_*/;
    public /*_virtual_*/ native NdbTransaction/*_NdbTransaction *_*/ getNdbTransaction() /*_const_*/;
    public final native NdbLockHandle/*_const NdbLockHandle *_*/ getLockHandle() /*_const_*/;
    public final native NdbLockHandle/*_const NdbLockHandle *_*/ getLockHandleM();
    public /*_virtual_*/ native int insertTuple();
    public /*_virtual_*/ native int updateTuple();
    public /*_virtual_*/ native int writeTuple();
    public /*_virtual_*/ native int deleteTuple();
    public /*_virtual_*/ native int readTuple(int/*_LockMode_*/ p0);
    public final native int equal(String/*_const char *_*/ anAttrName, ByteBuffer/*_const char *_*/ aValue);
    public final native int equal(String/*_const char *_*/ anAttrName, int/*_Int32_*/ aValue); // MMM covers signed and unsigned integral types
    public final native int equal(String/*_const char *_*/ anAttrName, long/*_Int64_*/ aValue); // MMM covers signed and unsigned integral types
    public final native int equal(int/*_Uint32_*/ anAttrId, ByteBuffer/*_const char *_*/ aValue);
    public final native int equal(int/*_Uint32_*/ anAttrId, int/*_Int32_*/ aValue); // MMM covers signed and unsigned integral types
    public final native int equal(int/*_Uint32_*/ anAttrId, long/*_Int64_*/ aValue); // MMM covers signed and unsigned integral types
    public final native NdbRecAttr/*_NdbRecAttr *_*/ getValue(String/*_const char *_*/ anAttrName, ByteBuffer/*_char *_*/ aValue /*_= 0_*/);
    public final native NdbRecAttr/*_NdbRecAttr *_*/ getValue(int/*_Uint32_*/ anAttrId, ByteBuffer/*_char *_*/ aValue /*_= 0_*/);
    public final native NdbRecAttr/*_NdbRecAttr *_*/ getValue(NdbDictionary.ColumnConst/*_const NdbDictionary.Column *_*/ p0, ByteBuffer/*_char *_*/ val /*_= 0_*/);
    public final native int setValue(String/*_const char *_*/ anAttrName, ByteBuffer/*_const char *_*/ aValue);
    public final native int setValue(String/*_const char *_*/ anAttrName, int/*_Int32_*/ aValue); // MMM covers signed and unsigned integral types
    public final native int setValue(String/*_const char *_*/ anAttrName, long/*_Int64_*/ aValue); // MMM covers signed and unsigned integral types
    public final native int setValue(String/*_const char *_*/ anAttrName, float aValue);
    public final native int setValue(String/*_const char *_*/ anAttrName, double aValue);
    public final native int setValue(int/*_Uint32_*/ anAttrId, ByteBuffer/*_const char *_*/ aValue);
    public final native int setValue(int/*_Uint32_*/ anAttrId, int/*_Int32_*/ aValue); // MMM covers signed and unsigned integral types
    public final native int setValue(int/*_Uint32_*/ anAttrId, long/*_Int64_*/ aValue); // MMM covers signed and unsigned integral types
    public final native int setValue(int/*_Uint32_*/ anAttrId, float aValue);
    public final native int setValue(int/*_Uint32_*/ anAttrId, double aValue);
    public /*_virtual_*/ native NdbBlob/*_NdbBlob *_*/ getBlobHandleM(String/*_const char *_*/ anAttrName); // MMM renamed due to nameclash with const version
    public /*_virtual_*/ native NdbBlob/*_NdbBlob *_*/ getBlobHandleM(int/*_Uint32_*/ anAttrId); // MMM renamed due to nameclash with const version
    public final native int setAbortOption(int/*_AbortOption_*/ p0);
    public interface /*_struct_*/ GetValueSpecConst
    {
        NdbDictionary.ColumnConst/*_const NdbDictionary.Column *_*/ column();
        // MMM! support <out:BB> or check if needed: ByteBuffer/*_void *_*/ appStorage();
        NdbRecAttr/*_NdbRecAttr *_*/ recAttr();
    }
    static public interface GetValueSpecConstArray extends ArrayWrapper< GetValueSpecConst >
    {
    }
    static public class GetValueSpecArray extends Wrapper implements GetValueSpecConstArray
    {
        static public native GetValueSpecArray create(int length);
        static public native void delete(GetValueSpecArray e);
        public native GetValueSpec at(int i);
    }
    static public class /*_struct_*/ GetValueSpec extends Wrapper implements GetValueSpecConst
    {
        public final native NdbDictionary.ColumnConst/*_const NdbDictionary.Column *_*/ column();
        // MMM! support <out:BB> or check if needed: public final native ByteBuffer/*_void *_*/ appStorage();
        public final native NdbRecAttr/*_NdbRecAttr *_*/ recAttr();
        public final native void column(NdbDictionary.ColumnConst/*_const NdbDictionary.Column *_*/ p0);
        // MMM! support <out:BB> or check if needed: public final native void appStorage(ByteBuffer/*_void *_*/ p0);
        public final native void recAttr(NdbRecAttr/*_NdbRecAttr *_*/ p0);
        static public final native GetValueSpec create();
        static public final native void delete(GetValueSpec p0);
    }
    static public interface SetValueSpecConstArray extends ArrayWrapper< SetValueSpecConst >
    {
    }
    static public class SetValueSpecArray extends Wrapper implements SetValueSpecConstArray
    {
        static public native SetValueSpecArray create(int length);
        static public native void delete(SetValueSpecArray e);
        public native SetValueSpec at(int i);
    }
    public interface /*_struct_*/ SetValueSpecConst
    {
        NdbDictionary.ColumnConst/*_const NdbDictionary.Column *_*/ column();
        // MMM! support <out:BB> or check if needed: ByteBuffer/*_const void *_*/ value();
    }
    static public class /*_struct_*/ SetValueSpec extends Wrapper implements SetValueSpecConst
    {
        public final native NdbDictionary.ColumnConst/*_const NdbDictionary.Column *_*/ column();
        // MMM! support <out:BB> or check if needed: public final native ByteBuffer/*_const void *_*/ value();
        public final native void column(NdbDictionary.ColumnConst/*_const NdbDictionary.Column *_*/ p0);
        // MMM! support <out:BB> or check if needed: public final native void value(ByteBuffer/*_const void *_*/ p0);
        static public final native SetValueSpec create();
        static public final native void delete(SetValueSpec p0);
    }
    public interface /*_struct_*/ OperationOptionsConst
    {
        long/*_Uint64_*/ optionsPresent();
        public interface /*_enum_*/ Flags
        {
            int OO_ABORTOPTION = 0x01,
                OO_GETVALUE = 0x02,
                OO_SETVALUE = 0x04,
                OO_PARTITION_ID = 0x08,
                OO_INTERPRETED = 0x10,
                OO_ANYVALUE = 0x20,
                OO_CUSTOMDATA = 0x40;
        }
        int/*_AbortOption_*/ abortOption();
        GetValueSpecArray/*_GetValueSpec *_*/ extraGetValues();
        int/*_Uint32_*/ numExtraGetValues();
        SetValueSpecConstArray/*_const SetValueSpec *_*/ extraSetValues();
        int/*_Uint32_*/ numExtraSetValues();
        int/*_Uint32_*/ partitionId();
        NdbInterpretedCodeConst/*_const NdbInterpretedCode *_*/ interpretedCode();
        int/*_Uint32_*/ anyValue();
        // MMM ByteBuffer/*_void *_*/ customData();
    }
    static public class /*_struct_*/ OperationOptions extends Wrapper implements OperationOptionsConst
    {
        static public final native int/*_Uint32_*/ size();
        public final native long/*_Uint64_*/ optionsPresent();
        public final native int/*_AbortOption_*/ abortOption();
        public final native GetValueSpecArray/*_GetValueSpec *_*/ extraGetValues();
        public final native int/*_Uint32_*/ numExtraGetValues();
        public final native SetValueSpecConstArray/*_const SetValueSpec *_*/ extraSetValues();
        public final native int/*_Uint32_*/ numExtraSetValues();
        public final native int/*_Uint32_*/ partitionId();
        public final native NdbInterpretedCodeConst/*_const NdbInterpretedCode *_*/ interpretedCode();
        public final native int/*_Uint32_*/ anyValue();
        // MMM! support <out:BB> or check if needed: public final native ByteBuffer/*_void *_*/ customData();
        public final native void optionsPresent(long/*_Uint64_*/ p0);
        public final native void abortOption(int/*_AbortOption_*/ p0);
        public final native void extraGetValues(GetValueSpecArray/*_GetValueSpec *_*/ p0);
        public final native void numExtraGetValues(int/*_Uint32_*/ p0);
        public final native void extraSetValues(SetValueSpecConstArray/*_const SetValueSpec *_*/ p0);
        public final native void numExtraSetValues(int/*_Uint32_*/ p0);
        public final native void partitionId(int/*_Uint32_*/ p0);
        public final native void interpretedCode(NdbInterpretedCodeConst/*_const NdbInterpretedCode *_*/ p0);
        public final native void anyValue(int/*_Uint32_*/ p0);
        // MMM! support <out:BB> or check if needed: public final native public final native void customData(ByteBuffer/*_void *_*/ p0);
        static public final native OperationOptions create();
        static public final native void delete(OperationOptions p0);
    }
}
